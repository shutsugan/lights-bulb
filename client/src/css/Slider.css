.Slider__container {
	position: sticky;
	top: 0px;
	right: 0;
	width: 400px;
	height: 100vh;
	opacity: .4;
	pointer-events: none;
}

.Slider--active {
	opacity: 1;
	pointer-events: auto;
}

.Slider {
	position: absolute;
	width: 400px;
	height: 400px;
	border: 1px solid transparent;
	border-radius: 0px 6px 6px 0;
	top: 0;
	right: 0;
	background-color: #1d1d25;
	-webkit-touch-callout: none;
	-webkit-user-drag: none;
	-webkit-user-select: none;
	padding: 0;
    margin: 0;
}

.Slider__value {
	position: absolute;
	left: 50%;
	margin-left: -50px;
	width: 100px;
	top: 160px;
	z-index: 100;
	font-size: 40px;
	font-weight: bold;
	font-family: 'Roboto', sans-serif;
	display: block;
	text-align: center;
	-webkit-touch-callout: none;
	-webkit-user-drag: none;
	-webkit-user-select: none;
	pointer-events: none;
	color: #FFF;
	text-shadow: 0px 1px 0px black;
}

.Slider__brightness {
	position: absolute;
	top: 52%;
	left: calc(50% - 42px);
	z-index: 10;
	font-size: 1.1em;
	font-family: 'Roboto', sans-serif;
	font-weight: 100;
	color: #FFF;
}

.Slider__icon{
	position: absolute;
	top: 34%;
	left: calc(50% - 7px);
	z-index: 10;
	color: #e0ac00;
	font-weight: bold;
}

.Slider__per {
	font-size: 12px;
}

.Slider__circle {
	position: relative;
	width: 300px;
	height: 300px;
	left: 50%;
	top: 50%;
	margin-left: -150px;
	margin-top: -150px;
	border-radius: 300px;
	box-shadow: 0 1px 1px 1px rgba(0, 0, 0, 0.4);
	background: #242329;
	cursor: pointer;
	-webkit-touch-callout: none;
	-webkit-user-drag: none;
	-webkit-user-select: none;
}

.Slider__inner-circle {
	position: absolute;
	width: 280px;
	height: 280px;
	top: 50%;
	left: 50%;
	margin-left: -140px;
	margin-top: -140px;
	border-radius: 290px;
	background-color: #1d1d25;
	background-position: center center;
	box-shadow: 0 1px 1px 1px rgba(0, 0, 0, 0.8);
	cursor: pointer;
	-webkit-touch-callout: none;
	-webkit-user-drag: none;
	-webkit-user-select: none;
}

.Slider__knob {
	box-sizing: border-box;
	position: absolute;
	top: -10px;
	left: 140px;
	width: 30px;
	height: 30px;
	text-align: center;
	border-radius: 20px;
	cursor: pointer;
	z-index: 10;
	background: #e0ac00;
	border: solid 4px #FFF;
	-webkit-touch-callout: none;
	-webkit-user-drag: none;
	-webkit-user-select: none;
}

.Slider__progress {
	position: absolute;
	top: 0;
	left: 0;
	width: 400px;
	height: 400px;
	-webkit-touch-callout: none;
	-webkit-user-drag: none;
	-user-select: none;
}

.arrow-left {
	position: absolute;
	top: 185px;
	left: -10px;
	width: 0; 
	height: 0; 
	border-top: 10px solid transparent;
	border-bottom: 10px solid transparent; 
	border-right:10px solid #1d1d25; 
}

@media only screen and (max-width: 860px) {
    .Main {
		flex-flow: column;
	}

	.arrow-left {
		border: none;
	}

	.Slider__container {
		width: 100%;
	    height: 416px;
	    position: relative;
	    display: flex;
	    justify-content: center;
	    align-items: center;
		background: #1d1d25;
		border-radius: 6px 6px 0 0;
	}

	.Slider {
		left: calc(50% - 200px) !important;
	}
}