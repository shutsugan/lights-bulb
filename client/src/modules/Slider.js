'use strict';

class Slider {
	/**
 	 * @param {HTMLElement} context the figure of the slider
 	 * @param {number} position brightness of the row
	 */
	constructor(context, position) {
		this._select = element => document.querySelector(element);
		this.context = context;
		this.position = position;

		this.raf = null;
		this.m_down = false;
		this.m_pos = {x: 0, y: 0};
		this.radius = 146;
		this.knob_radius = 15;
		this.max_diff = 150;
		this.constraint = 360;
		this.target = 0;

		this.x = 0;
		this.y = 0;

		this.circle = this.context.querySelector('.Slider__circle');
		this.inner_circle = this.context.querySelector('.Slider__inner-circle');
		this.knob = this.context.querySelector('.Slider__knob');
		this.progress = this.context.querySelector('.Slider__progress');
		this.ctx = this.progress.getContext('2d');
		this.value = this._select('.Slider__value');

		const circleOffset = this.circle.getBoundingClientRect();
	    const scrollLeft = window.pageXOffset || document.documentElement.scrollLeft;
	    const scrollTop = window.pageYOffset || document.documentElement.scrollTop;

	    this.element_position = {
      		x: circleOffset.left + scrollLeft,
      		y: circleOffset.top + scrollTop
	    };
		this.center_x = this.progress.offsetWidth / 2;
		this.center_y = this.progress.offsetHeight / 2;
		this.canvas_size = this.progress.offsetWidth;

		//calcular the brightness of the selected row
		this.target = (this.position / 100) * 360;
		const value = Math.round((this.target / 360) * 100);
		this.value.innerHTML = `${value}<span class="Slider__per"> %<span>`; 

		this._drawArc();
		this._addEventListener();
		this._updateSlider();
	}

	_addEventListener() {
		this.context.addEventListener('mousedown', event => this.mouseDown(event));
		this.context.addEventListener('mousemove', event => this.mouseMove(event));

		this._select('body').addEventListener('mouseup', event => this.mouseUp(event));
	}

	/**
 	 * update the position of the knob.
	 */
	_updateSlider() {
		this.raf = requestAnimationFrame(this._updateSlider.bind(this));
		this._setPosition();
	}

	_setPosition() {
		this.x = Math.round(this.radius * Math.sin(this.target * Math.PI / 180)) + this.radius - this.knob_radius + 4;
		this.y = Math.round(this.radius * -Math.cos(this.target * Math.PI / 180)) + this.radius - this.knob_radius + 4;
	
		this.knob.style.left = `${this.x}px`;
		this.knob.style.top = `${this.y}px`;
	}

	/**
 	 * draw the layers of the slider.
	 */
	_drawArc() {
		this.progress.width = this.canvas_size;
		this.progress.height = this.canvas_size;

		this.ctx.save();
		this.ctx.translate(this.center_x, this.center_y - this.radius);
		this.ctx.rotate(-90 * Math.PI / 180);

		const gradient = this.ctx.createLinearGradient(100,-50,50,160);
		gradient.addColorStop(0, '#e0ac00');
		gradient.addColorStop(1, '#71590a');//#b58b01

		this.ctx.strokeStyle = gradient;
		this.ctx.beginPath();
		this.ctx.lineWidth = 17;
		this.ctx.arc(0 - this.radius - 1, 0, this.radius - 1, 0, this.target * Math.PI / 180, false);
		this.ctx.stroke();
		this.ctx.restore();
	}

	/**
 	 * get the value to be displayed in the center of the slider.
 	 * @params {Event} event
	 */
	_setMousePosition(event) {
		this.m_pos = {
			x: event.pageX - this.element_position.x,
			y: event.pageY - this.element_position.y
		};

		const atan = Math.atan2(this.m_pos.x - this.radius, this.m_pos.y - this.radius);
		const target = -atan / (Math.PI / 180) + 180;
		const diff = Math.abs(target - this.target);

		if (diff < this.max_diff && target < this.constraint) {
			this.target = target;
			this._drawArc();

			let value = Math.round((this.target / 360) * 100);
			if (value > 100) value = 100;
			this.value.innerHTML = `${value}<span class="Slider__per"> %<span>`;
		}
	}

	mouseDown(event) {
		this.m_down = true;
	}

	mouseUp(event) {
		this.m_down = false;
	}

	mouseMove(event) {
		if (this.m_down) this._setMousePosition(event);
	}
}

export default Slider;