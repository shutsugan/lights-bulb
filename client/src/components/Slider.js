import React, {Component} from 'react';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faSun } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import SliderRange from '../modules/Slider';

import '../css/Slider.css';

library.add(faSun);

class Slider extends Component {
	constructor(props) {
		super(props);

		this.state = {
			brightness: null
		}
	}

	componentDidUpdate() {
		this._initSlider();
	}

	/**
 	 * Initialize the sliderb.
	 */
	_initSlider = _ => {
		const element = document.querySelector('.Slider');
		const brightness = this.props.current.brightness ? this.props.current.brightness : this.state.brightness;
		new SliderRange(element, brightness);
	}

	/**
 	 * Update the brightness value of the burrent light bulb.
 	 * @param {Event} event
	 */
	_setBrightness = event => {
		const value = document.querySelector('.Slider__value').textContent;
		const brightness = parseInt(value.split('%')[0]);
		const active = brightness === 0 ? false : true;

		const id = this.props.current.id;
		const data = {
			brightness: brightness,
			active: active
		}

		this._update(id, data);

		const current = {
			element: this.props.current.element,
			name: this.props.current.name,
			id: this.props.current.id,
			brightness: brightness
		};

		this.props._getCurrent(current);
	}

	/**
 	 * Update a device.
 	 * @param {number} id the id of the device to update
 	 * @param {object} data the payload of which the device will be changed by
	 */
	_update = (id, data) => {
		this.props._update(id, data);
	}

	render() {
		const active = this.props.current.brightness !== undefined ? 'Slider--active' : '';
		
		return(
			<div className={`Slider__container ${active}`}>
				<figure className="Slider" onMouseUp={this._setBrightness}>
					<span className="Slider__value">0</span>
					<span className="Slider__icon">
						<FontAwesomeIcon icon="sun" />
					</span>
					<span className="Slider__brightness">Brightness</span>
					<div className="Slider__circle">
						<div className="Slider__knob" unselectable="on"></div>
					</div>
					<canvas className="Slider__progress"></canvas>
					<div className="Slider__inner-circle"></div>
				</figure>
				<div className="arrow-left"></div>
			</div>
		);
	}
}

export default Slider;