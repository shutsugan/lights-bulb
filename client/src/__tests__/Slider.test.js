import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import Slider from '../components/Slider';

const device = {
	brightness: 0
};

describe('Slider', () => {
    it('should be defined', () => {
      expect(Slider).toBeDefined();
    });
  
    it('should render correctly', () => {
      const tree = shallow(<Slider current={device} />);
      expect(tree).toMatchSnapshot();
    });
});