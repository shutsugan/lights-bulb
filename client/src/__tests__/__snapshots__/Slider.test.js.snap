// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="Slider__container Slider--active"
>
    <figure
        className="Slider"
        onMouseUp={[Function]}
    >
        <span
            className="Slider__value"
        >
            0
        </span>
        <span
            className="Slider__icon"
        >
            <FontAwesomeIcon
                border={false}
                className=""
                fixedWidth={false}
                flip={null}
                icon="sun"
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                transform={null}
            />
        </span>
        <span
            className="Slider__brightness"
        >
            Brightness
        </span>
        <div
            className="Slider__circle"
        >
            <div
                className="Slider__knob"
                unselectable="on"
            />
        </div>
        <canvas
            className="Slider__progress"
        />
        <div
            className="Slider__inner-circle"
        />
    </figure>
    <div
        className="arrow-left"
    />
</div>,
  "nodes": Array [
    <div
      className="Slider__container Slider--active"
>
      <figure
            className="Slider"
            onMouseUp={[Function]}
      >
            <span
                  className="Slider__value"
            >
                  0
            </span>
            <span
                  className="Slider__icon"
            >
                  <FontAwesomeIcon
                        border={false}
                        className=""
                        fixedWidth={false}
                        flip={null}
                        icon="sun"
                        inverse={false}
                        listItem={false}
                        mask={null}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        symbol={false}
                        transform={null}
                  />
            </span>
            <span
                  className="Slider__brightness"
            >
                  Brightness
            </span>
            <div
                  className="Slider__circle"
            >
                  <div
                        className="Slider__knob"
                        unselectable="on"
                  />
            </div>
            <canvas
                  className="Slider__progress"
            />
            <div
                  className="Slider__inner-circle"
            />
      </figure>
      <div
            className="arrow-left"
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_context": Object {},
    "_element": <Slider
      current={
            Object {
                  "brightness": 0,
                }
      }
/>,
    "_instance": Slider {
      "_initSlider": [Function],
      "_setBrightness": [Function],
      "_update": [Function],
      "context": Object {},
      "props": Object {
        "current": Object {
          "brightness": 0,
        },
      },
      "refs": Object {},
      "state": Object {
        "brightness": null,
      },
      "updater": Updater {
        "_renderer": [Circular],
      },
    },
    "_newState": null,
    "_rendered": <div
      className="Slider__container Slider--active"
>
      <figure
            className="Slider"
            onMouseUp={[Function]}
      >
            <span
                  className="Slider__value"
            >
                  0
            </span>
            <span
                  className="Slider__icon"
            >
                  <FontAwesomeIcon
                        border={false}
                        className=""
                        fixedWidth={false}
                        flip={null}
                        icon="sun"
                        inverse={false}
                        listItem={false}
                        mask={null}
                        pull={null}
                        pulse={false}
                        rotation={null}
                        size={null}
                        spin={false}
                        symbol={false}
                        transform={null}
                  />
            </span>
            <span
                  className="Slider__brightness"
            >
                  Brightness
            </span>
            <div
                  className="Slider__circle"
            >
                  <div
                        className="Slider__knob"
                        unselectable="on"
                  />
            </div>
            <canvas
                  className="Slider__progress"
            />
            <div
                  className="Slider__inner-circle"
            />
      </figure>
      <div
            className="arrow-left"
      />
</div>,
    "_rendering": false,
    "_updater": Updater {
      "_renderer": [Circular],
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Slider
    current={
        Object {
            "brightness": 0,
          }
    }
/>,
}
`;
